pytorch_lightning/plugins/training_type/ddp.py

```diff
@@ -373,3 +373,7 @@ def register_plugins(cls, plugin_registry: Dict) -> None:
             description="DDP Plugin with `find_unused_parameters` as False",
             find_unused_parameters=False
         )
+
+    def __del__(self) -> None:
+        if torch_distrib.is_initialized():
+            torch_distrib.destroy_process_group()
```

Comment:
@awaelchli do you think we should still call `self.teardown()` here?

which is the parent's:

```python
        if self.on_gpu:
            # GPU teardown
            self.lightning_module.cpu()
            # clean up memory
            with torch.cuda.device(self.root_device):
                torch.cuda.empty_cache()
```

maybe just the empty cache?

Reply:
I don't know if process group leaves memory in cache. probably it does. 
but I think teardown generally should only run once per fit. so we might just want to call empty_cache() in del.

Reply:
```suggestion
    def __del__(self) -> None:
        if torch_distrib.is_initialized():
            torch_distrib.destroy_process_group()
            # clean up memory
            with torch.cuda.device(self.root_device):
                torch.cuda.empty_cache()
```

Reply:
I think we should do this (although I think when the process group is destroyed, it's GPU memory should also be freed since there is nothing holding it anymore)

Reply:
n00b question: why do the process group destruction here vs in the training type plugin's teardown?

Reply:
> why do the process group destruction in the plugin's destructor vs vs in the training type plugin's teardown?

Quoting Adrian:

> but we don’t want to destroy the group after fit. it needs to stay alive until the process ends.
> we want to run more after the first Trainer.fit(), like Trainer.test() validate() or predict()

> if someone instantiates a plugin object and passes it into the trainer, it's still live after the trainer.fit call, so the process group won't be destroyed right? then wouldn't subsequent calls to the trainer could still observe the issue mentioned in the PR summary?

This would only be an issue when the second trainer call uses a different distributed process group configuration to the first trainer call. This would be an extreme edge case for users. But it is not on our CI where pytests runs the tests run on the same process